feat(lep): multiple backup stores support (#6630)
longhorn/longhorn: enhancements/20230828-multiple-backup-stores-support.md

# Multiple Backup Stores Support

## Summary

> The current Longhorn only support a single backup store for all volumes. The feature aims to provide multiple backup stores with tiering.

Reviewer: Actually, we don't support tiering in the implementation. `The feature aims to provide multiple backup stores` is good enough.

Reviewer 2: @derekbit what tiering (use cases) you think we could/should support in the future?

Author: Removed `with tiering`

### Related Issues

https://github.com/longhorn/longhorn/issues/2317
https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup stores to that they want to back up volumes.

### Goals

- There are difference backup targets at the same time on Longhorn.

> - Backups can be synchronized from difference remote backup targets.

Reviewer: ```suggestion
- Backups can be synchronized from different remote backup targets.
```

Author: Modified. Thanks

> - Backups can be backed up to difference remote backup targets.

Reviewer: ```suggestion
- Snapshots can be backed up to different remote backup targets.
```

Author: Modified. Thanks

> - A backup can be restored from difference remote backup targets to be a new volume.

Reviewer: ```suggestion
- A backup can be restored from its remote backup targets to be a new volume.
```

Author: Modified. Thanks

### Non-goals [optional]

`None`

## Proposal

1. Allow users to create, delete or modify extra backup targets except `default`.
     - backup target controller needs to synchronize the data from the extra backup target created.
     - backup target controller needs to delete the data in the cluster related the extra backup target deleted.
2. Introduce new fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status` of both BackupVolume and Backup CRDs to keep backup target information.

### User Stories

User can setup the default backup target as usual on the page `Setting/General` and create extra backup targets on the page `Settings/Backup Target` or create/modify backup targets by the manifest. After extra backup targets created and synchronized, users can start to create a backup on these remote backup targets.

### User Experience In Detail

#### Create Or Modify Backup Targets By UI

> 1. User can find the link `Backup Target` in drop down menu of `Setting`.

Reviewer: We probably can remove the `backup-target` and `backup-target-credential-secret` from the settings. One of the backup targets is set to the default one.

Author: Added in `Implementation Overview`>`CRD` to mention that we will remove these two settings.

2. We have a default backup target on the page `Backup Target` and it would show its name, backup target url, secret name for this backup target (credential), poll interval and availability status.
3. User can create a new backup target on the page `Backup Target` by clicking the button `Create`.
4. User have to fill out the necessary items for backup target name and url, and secret(credential) name, poll interval can be optional.
5. User can click the button `Edit` to modify the backup target information.

#### Create Or Modify Backup Targets By CLI

User can the bash command to create or modify a backup target such as below:

```bash
cat <<EOF >>new-backup-target.yaml
apiVersion: longhorn.io/v1beta2
kind: BackupTarget
metadata:
  name: azure-blob-server-001
  namespace: longhorn-system
spec:
  backupTargetURL: azblob://demo@core.windows.net/
  credentialSecret: azblob-secret
  pollInterval: 4m30s
status:

>   default: true

Reviewer: Can user delete the default backup target?

Author: I think it would be like clean up the backup target URL of default backup target we implement now, so default backup target can be deleted: 

1. if it is the only one backup target.
2. if there are other backup targets, and then we choose another one as default randomly.

WDYT?

    | Delete | name string | err error | Remove a backup target and its related backup volume and backup CR objects in the cluster | **DELETE** `/v1/backuptargets/{backupTargetName}` |
    | Get    |  | backupTarget BackupTarget, err error | Get the backup targets information | **GET** `/v1/backuptargets/{backupTargetName}` |
    | Update | backupTargetURL, credentialSecret string, pollInterval time.Duration, default boolean | err error | Update the backup targets information | **PUT** `/v1/backuptargets/{backupTargetName}` |

- Modify the APIs `BackupVolumeGet`, `BackupVolumeList`:
  - Add new fields `BackupTargetName` and `VolumeName` in returning `BackupVolume` information.
- Modify the APIs `BackupCreate`, `BackupGet`:
  - Add a new field `BackupTargetName` in returning `Backup` information.

## Design

### Implementation Overview

#### CRDs

1. Modify the Backup CRD `backups.longhorn.io` to add new fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status`.

  ```yaml
  metadata:
    name: the backup name. (string)
  spec:
    backupTargetURL: the backup target URL. (string)
  status:

>     backupTargetName: the backup target Name. (string)

Reviewer: What's the purpose of the `status.backupTargetName`?

Author: It is used to get the BackupTarget CR conveniently

Reviewer 3: This has backupTargetName twice.

Author: Removed. Thanks

# Multiple Backup Stores Support

## Summary

The current Longhorn only support a single backup store for all volumes. The feature aims to provide multiple backup stores.

### Related Issues

https://github.com/longhorn/longhorn/issues/2317
https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup stores to that they want to back up volumes.

### Goals

- There are different backup targets at the same time on Longhorn.
- Backups can be synchronized from different remote backup targets.
- Snapshots can be backed up to different remote backup targets.
- A backup can be restored from its remote backup targets to be a new volume.

### Non-goals [optional]

`None`

## Proposal

1. Allow users to create, delete or modify extra backup targets except `default`.

>      - backup target controller needs to synchronize the data from the extra backup target created.

Reviewer 3: What data are synchronized, and to where?

Author: Modified. The backups information will be synchronized from remote backup target and controller will create Backup/BackupVolume CR objects that exists on remote backup target.

>      - backup target controller needs to delete the data in the cluster related the extra backup target deleted.

Reviewer 3: What does this mean?  When a backup target is deleted, all the backups on it are implicitly deleted?  That seems extreme.  Shouldn't we require that there be no backups stored there before we delete a backup target?

Another question:  does it make sense to have a "read-only" status on a backup target such that one can restore existing backups, but not use it to make new ones?

Author: > What does this mean? When a backup target is deleted, all the backups on it are implicitly deleted?

Only Backup/BackupVolume CR objects related to the deleted backup target in the cluster will be deleted and data on remote backup target will not be deleted as now we do for `BackupTargets.Spec.BackupTargetURL` is empty.

Modified the description to be more clear.

> does it make sense to have a "read-only" status on a backup target such that one can restore existing backups, but not use it to make new ones?

Yes, it is a good idea. 
Thanks.

2. Introduce new fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status` of both BackupVolume and Backup CRDs to keep backup target information.

### User Stories

> User can setup the default backup target as usual on the page `Setting/General` and create extra backup targets on the page `Settings/Backup Target` or create/modify backup targets by the manifest. After extra backup targets created and synchronized, users can start to create a backup on these remote backup targets.

Reviewer 3: When backing up to multiple targets, is each backup independent (duplicating the process of calculating diffs and so on?)  Or is there one actual backup operation, and each write is "mirrored" to the set of targets as we go?

Author: >When backing up to multiple targets, is each backup independent (duplicating the process of calculating diffs and so on?)

Yes, it is independent.

4. User have to fill out the necessary items for backup target name and url, and secret(credential) name, poll interval can be optional.
5. User can click the button `Edit` to modify the backup target information.

#### Create Or Modify Backup Targets By CLI

User can the bash command to create or modify a backup target such as below:

```bash
cat <<EOF >>new-backup-target.yaml
apiVersion: longhorn.io/v1beta2
kind: BackupTarget
metadata:
  name: azure-blob-server-001
  namespace: longhorn-system
spec:
  backupTargetURL: azblob://demo@core.windows.net/
  credentialSecret: azblob-secret
  pollInterval: 4m30s
status:
  default: true
EOF

kubectl apply -f new-backup-target.yaml
```

#### Create A Backup To Remote Backup Targets By UI

After users click the button `Create Backup` on the `Volume Details` page, they can choose multiple backup targets they just created by checkboxes to create backups on backup targets.

#### Create A Backup To Remote Backup Targets By CLI

1. Create a snapshot of the volume first
2. Create backups for each backup target you want to store the snapshot by the manifest as below:

> ```txt

Reviewer 3: ```yaml

Author: Modified. Thanks

  name: backup-test-001
  namespace: longhorn-system
spec:
  backupTarget: azure-blob-server-001
  labels:
    longhorn.io/volume-access-mode: rwo
  snapshotName: xxxxxxxx-9a69-4ede-ab88-c9853459462c
status:
  default: true
```

#### Restore From A Backup From A Remote Backup Target

Restoring from a backup will behave as before. Choose a backup and do the `Restore` operation in drop down menu on UI.

### API changes

- Introduce new APIs `BackupTargetCreate`,`BackupTargetDelete` and `BackupTargetGet`:

    | API | Input | Output | Comments | HTTP Endpoint |
    | --- | --- | --- | --- | --- |
    | Create | name, backupTargetURL, credentialSecret string, pollInterval time.Duration, default boolean | err error | Create a new backup target and start to synchronize data | **POST** `/v1/backuptargets/{backupTargetName}` |
    | Delete | name string | err error | Remove a backup target and its related backup volume and backup CR objects in the cluster | **DELETE** `/v1/backuptargets/{backupTargetName}` |
    | Get    |  | backupTarget BackupTarget, err error | Get the backup targets information | **GET** `/v1/backuptargets/{backupTargetName}` |

>     | Update | backupTargetURL, credentialSecret string, pollInterval time.Duration, default boolean | err error | Update the backup targets information | **PUT** `/v1/backuptargets/{backupTargetName}` |

Reviewer 3: If one modifies a backup target to be the default one, is the attribute automatically removed from the previous default in a single transaction?  If not, does it validate that there is not currently a default?

Author: >If one modifies a backup target to be the default one, is the attribute automatically removed from the previous default in a single transaction?

Yes, it is.
Added more descriptions for this in `User Experience In Detail` and `Implementation Overview`
Thanks

  spec:
    backupTargetName: the backup target Name. (string)
    backupTargetURL: the backup target URL. (string)
    backupTargetName: the backup target Name. (string)
  ```

2. Modify the Backup Volume CRD `backupvolumes.longhorn.io` to add new fields `BackupTargetURL` and `VolumeName` in `Spec` and `BackupTargetName` in `Status`.

  ```yaml
  metadata:
    name: the backup name. (string)
  spec:
    backupTargetName: the backup target Name. (string)
    backupTargetURL: the backup target URL. (string)
    volumeName:  the volume name for backups. (string)
  ```

3. Modify the Backup Target CRD `backuptagets.longhorn.io` to add a new fields `Default` in `Status`.

  ```yaml
  metadata:
    name: the backup name. (string)
  ...
  status:
    default: the backup target is default or not. (boolean)

4. Remove `backup-target` and `backup-target-credential-secret` from the settings.
  ```

#### Backup Related Controllers

1. Remove `default` backup target from the setting controller.

2. Modify backup target controller to allow to add and delete a extra backup target.

   - Clean up backup volume CR objects in the cluster by label `backup-target` when deleting a backup target.

>    - Create a pulling backup volume from the backup target by the backup volume name `volumeName` + `-` + `backupTarget.Name` and adding a label `backup-target` with the backup target URL.

Reviewer 3: What is a "pulling" backup volume?  What is it pulling, and from where?

Author: Modified the description. A `pulling` backup volume is a backup volume existing on the remote backup target and we prepare to create a BackupVolume CR object for it when we synchronize the backup information from the remote backup target.

   - Get backup volume CR objects in the cluster by label `backup-target` with the backup target URL.

3. Modify backup volume controller to tell which backup volume belongs to which backup target and synchronize the backups of the backup volume from remote backup target.

   - Clean up backup CR objects in the cluster by the volume name `backupVolume.Spec.volumeName` and `backupVolume.Spec.backupTargetURL` when deleting a backup volume.
   - Create a pulling backup from the backup target by the volume name `backupVolume.Spec.volumeName` and `backupVolume.Spec.backupTargetURL`.

### Test plan

1. Setup the default backup target.
2. Create a backup to default backup target and it succeeds.

> 3. Setup an extra old backup target.

Reviewer 3: What makes it "old"?

4. Backups created with default backup target will not be deleted.
5. Backups on the extra old backup target can be synchronized.
6. Create a backup to extra backup target and it succeeds.
7. Restore a backup from default remote backup target, and it succeeds and data is correct.
8. Restore a backup from the extra remote backup target, and it succeeds and data is correct.

### Upgrade strategy

1. Fill in fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status` for backup CR objects with corresponding values.
2. Fill in fields `BackupTargetURL` and `VolumeName` in `Spec` and `BackupTargetName` in `Status` for backup volume CR objects with corresponding values.

> 

Reviewer 3: How is the migration from a global setting to a new CRD done during upgrade?

Author: We have the `default` backup target created now when the setting controller is running.
We just need to fill in the values of `default` backup target during upgrade.

# Multiple Backup Stores Support

## Summary

The current Longhorn only support a single backup store for all volumes. The feature aims to provide multiple backup stores.

### Related Issues

https://github.com/longhorn/longhorn/issues/2317
https://github.com/longhorn/longhorn/issues/5411

## Motivation

Users can choose backup stores to that they want to back up volumes.

### Goals

- There are different backup targets at the same time on Longhorn.
- Backups can be synchronized from different remote backup targets.
- Snapshots can be backed up to different remote backup targets.
- A backup can be restored from its remote backup targets to be a new volume.

### Non-goals [optional]

`None`

> ## Proposal

Reviewer: What will happen when two backup targets have the same backup volumes?

Author: The backup volume CR object name will be `volume name + backup target name as a postfix concatenated`, so we add fields `backupTargetName`, `backupTargetURL` and `volumeName` to tell which backup volume is belong to which volume in the cluster and which remote backup target.

1. Allow users to create, delete or modify backup targets.
     - backup target controller needs to synchronize the backups information from the remote backup target and create Backup CR objects and BackupVolume CR objects in the cluster when the backup target is created.
     - When the backup target is deleted, backup target controller needs to delete the Backup CR and BackupVolume CR objects in the cluster related to the backup target.
2. Introduce new fields `BackupTargetURL` and `BackupTargetName` in `Spec` of both BackupVolume and Backup CRDs to keep backup target information.

### User Stories

User need to create/modify backup targets on the page `Settings/Backup Target` or by the manifest. After backup targets are created and synchronized, users can start to create a backup on these remote backup targets.

### User Experience In Detail

#### Create Or Modify Backup Targets By UI

1. User can find the link `Backup Target` in drop down menu of `Setting`.
2. The page `Backup Target` would show backup target name, backup target url, secret name for this backup target (credential), poll interval, default and availability status.

> 3. User can create a new backup target on the page `Backup Target` by clicking the button `Create` and first backup target created will be the default backup target.

Reviewer: Can we create secrets on the UI for better UX?

4. User have to fill out the necessary items for backup target name and url, and secret(credential) name, poll interval and default can be optional.
5. User can click the button `Edit` to modify a backup target information.
6. User can choose another backup target to be a default backup target and previous default backup target will become a common backup target automatically.
7. User can click the button `Delete` to delete a backup target.
8. User can not delete or disable a default backup target directly.

#### Create Or Modify Backup Targets By CLI

User can the bash command to create or modify a backup target such as below:

```bash
cat <<EOF >>new-backup-target.yaml
apiVersion: longhorn.io/v1beta2

> kind: BackupTarget

Reviewer: Can we add `labels` in BackupTarget spec? Then, users can add some custom information.

Author: Modified to add labels in BackupTarget spec. Thanks

metadata:
  name: azure-blob-server-001
  namespace: longhorn-system
spec:
  backupTargetURL: azblob://demo@core.windows.net/
  credentialSecret: azblob-secret
  pollInterval: 4m30s

> status:

Reviewer: Can we introduce `status.conditions` indicating the status of the backup targets such as `connection status`?
For example, node resources' conditions are
```
    conditions:
    - lastProbeTime: ""
      lastTransitionTime: "2023-09-06T15:40:50Z"
      message: Node rancher70-worker1 is ready
      reason: ""
      status: "True"
      type: Ready
    - lastProbeTime: ""
      lastTransitionTime: "2023-09-06T15:40:50Z"
      message: ""
      reason: ""
      status: "True"
      type: Schedulable
    - lastProbeTime: ""
      lastTransitionTime: "2023-09-06T15:40:50Z"
      message: ""
      reason: ""
      status: "True"
      type: MountPropagation
``

Author: We have had the `status.conditions` in `BackupTargetStatus` of BackupTarget CRD.

    | API | Input | Output | Comments | HTTP Endpoint |
    | --- | --- | --- | --- | --- |
    | Create | name, backupTargetURL, credentialSecret string, pollInterval time.Duration, default boolean | err error | Create a new backup target and start to synchronize data | **POST** `/v1/backuptargets/{backupTargetName}` |
    | Delete | name string | err error | Remove a backup target and its related backup volume and backup CR objects in the cluster | **DELETE** `/v1/backuptargets/{backupTargetName}` |
    | Get    |  | backupTarget BackupTarget, err error | Get the backup targets information | **GET** `/v1/backuptargets/{backupTargetName}` |
    | Update | backupTargetURL, credentialSecret string, pollInterval time.Duration, default boolean | err error | Update the backup targets information | **PUT** `/v1/backuptargets/{backupTargetName}` |

- Modify the APIs `BackupVolumeGet`, `BackupVolumeList`:
  - Add new fields `BackupTargetName` and `VolumeName` in returning `BackupVolume` information.
- Modify the APIs `BackupCreate`, `BackupGet`:
  - Add a new field `BackupTargetName` in returning `Backup` information.

## Design

### Implementation Overview

#### CRDs

> 1. Modify the Backup CRD `backups.longhorn.io` to add new fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status`.

Author: Yes, both added when creating Backup and BackupVolume CR objects. Thanks

### User Stories

User need to create/modify backup targets on the page `Settings/Backup Target` or by the manifest. After backup targets are created and synchronized, users can start to create a backup on these remote backup targets.

### User Experience In Detail

#### Create Or Modify Backup Targets By UI

1. User can find the link `Backup Target Credential` in drop down menu of `Setting`.
2. The page `Backup Target Credential` would show the credential name, backup target type(such as s3, nfs, cifs and ablob), account name, account key and other parameters.
3. User can create a new secret(credential) on the page `Backup Target Credential` by clicking the button `Create`.
4. User can modify a new secret(credential) on the page `Backup Target Credential` by clicking the button `Edit`
5. User can delete a new secret(credential) on the page `Backup Target Credential` by clicking the button `Delete`
6. User can find the link `Backup Target` in drop down menu of `Setting`.
7. The page `Backup Target` would show backup target name, backup target url, secret name for this backup target (credential), poll interval, default and availability status.

> 8. User can create a new backup target on the page `Backup Target` by clicking the button `Create` and first backup target created will be the default backup target. User have to fill out the necessary items for backup target name and url, and secret name(listing the items in `Backup Target Credential`), poll interval, default and readOnly can be optional.

Author: >Is it legal to create another target with a different name but exactly the same URL and credentials?

I believe that it is not legal to create another target with a different name but exactly the same URL and credentials even though it might work well on the data engine side, it might be confused on the control plane side for handling backup volume and its backups on the same URL path.

>Is it legal to create another target with a different name but exactly the same URL host+path (although perhaps different mount options or other query tags)?

If it had exactly the same URL host+path w/wo different mount options, it is not legal as below.

@james-munson Thanks a lot, it is what I did not consider and handle.

> 8. User can create a new backup target on the page `Backup Target` by clicking the button `Create` and first backup target created will be the default backup target. User have to fill out the necessary items for backup target name and url, and secret name(listing the items in `Backup Target Credential`), poll interval, default and readOnly can be optional.

Author: Added a description of this discussion.

> 9. User can modify a backup target information by clicking the button `Edit`.

Author: >Is it legal to modify the name? (probably not, since existing backups have that name as a reference).

Yeah, the name should not be modified.

>Is it legal to modify the URL, keeping the same name?

It is legal as what we are doing now, we could modify the `SettingNameBackupTarget` `backup-target` and old backups would be removed then pull new backups on the new URL backup target.

>Is it legal to modify the URL to be the same as another existing target?

No, I believe it is not legal now for different backup targets with the same URL as mentioned in https://github.com/longhorn/longhorn/pull/6630#discussion_r1363013849.

> 9. User can modify a backup target information by clicking the button `Edit`.

Author: Added a description of this discussion.

    | Get    |  | secret SecretInput, err error | Get the secret information | **GET** `/v1/secrets/{secretName}` |

  ```golang
    type SecretInput struct {
      client.Resource

      Name       string            `json:"name"`
      SecretType string            `json:"secretType"`
      Data       map[string]string `json:"data"`
    }
  ```

- Modify the APIs `BackupVolumeGet`, `BackupVolumeList`:
  - Add new fields `BackupTargetName` and `VolumeName` in returning `BackupVolume` information.
- Modify the APIs `BackupCreate`, `BackupGet`:
  - Add a new field `BackupTargetName` in returning `Backup` information.
- Modify the API `SnapshotBackup`:
  - Add a new field `BackupTargetName` in the input `SnapshotInput`.

## Design

### Implementation Overview

#### CRDs

1. Modify the Backup CRD `backups.longhorn.io` to add new fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status`. And add a label LonghornLabelBackupTarget `backup-target` for resource listing.

  ```yaml
  metadata:
    labels:
      backup-target: the backup target name (string)
    name: the backup name. (string)
  spec:
    backupTargetName: the backup target Name. (string)

>     backupTargetURL: the backup target URL. (string)

Reviewer 4: I am wondering how do we handle the case user change the backup target's URL to a new URL. How do we handle the situation that the backup.Spec.backupTargetURL still pointing to old URL?

Author: It would be like the old way we handle the backup target URL changed, the old backups (CR objects) that do not exist on the new remote backup target would be removed from the cluster.

Author: Add a description of this behavior of the backup target controller.

    | Delete | name string | err error | Remove a secret | **DELETE** `/v1/secret/{secretName}` |
    | Update | type string, data map[string][]byte{} | err error | Update the secret information | **PUT** `/v1/secrets/{secretName}` |
    | Get    |  | secret []SecretInput, err error | Read a list of Secret | **GET** `/v1/secrets/` |
    | Get    |  | secret SecretInput, err error | Get the secret information | **GET** `/v1/secrets/{secretName}` |

  ```golang
    type SecretInput struct {
      client.Resource

      Name       string            `json:"name"`
      SecretType string            `json:"secretType"`
      Data       map[string]string `json:"data"`
    }
  ```

- Modify the APIs `BackupVolumeGet`, `BackupVolumeList`:
  - Add new fields `BackupTargetName` and `VolumeName` in returning `BackupVolume` information.
- Modify the APIs `BackupCreate`, `BackupGet`:
  - Add a new field `BackupTargetName` in returning `Backup` information.
- Modify the API `SnapshotBackup`:
  - Add a new field `BackupTargetName` in the input `SnapshotInput`.

## Design

### Implementation Overview

#### CRDs

> 1. Modify the Backup CRD `backups.longhorn.io` to add new fields `BackupTargetURL` in `Spec` and `BackupTargetName` in `Status`. And add a label LonghornLabelBackupTarget `backup-target` for resource listing.

Author: Yes, you are right. `BackupTargetName` should be in `Spec`.
Modified. Thanks.

  ```yaml
  metadata:
    labels:
      backup-target: the backup target name (string)
    name: the backup name. (string)
  spec:
    backupTargetName: the backup target Name. (string)
    backupTargetURL: the backup target URL. (string)
  ```

> 2. Modify the Backup Volume CRD `backupvolumes.longhorn.io` to add new fields `BackupTargetURL` and `VolumeName` in `Spec` and `BackupTargetName` in `Status`. And add a label LonghornLabelBackupTarget `backup-target` for resource listing.

Author: Modified. Thanks.
